grrrRRRRRRRRR!!!

```
19:25:33 [ERROR] Waited 600 seconds, no response
19:25:33 Kill the docker 67273fa384287e2fd59635526078ee0a3c08016e5837531122abec0fddac6724
```

--- a/lib/private/DB/Migrator.php
+++ b/lib/private/DB/Migrator.php
@@ -98,34 +98,6 @@ class Migrator {
 	}
 
 	/**
-	 * @param Schema $targetSchema
-	 * @throws \OC\DB\MigrationException
-	 */
-	public function checkMigrate(Schema $targetSchema) {
-		$this->noEmit = true;
-		/**@var \Doctrine\DBAL\Schema\Table[] $tables */
-		$tables = $targetSchema->getTables();
-		$filterExpression = $this->getFilterExpression();
-		$this->connection->getConfiguration()->
-			setFilterSchemaAssetsExpression($filterExpression);
-		$existingTables = $this->connection->getSchemaManager()->listTableNames();
-
-		$step = 0;
-		foreach ($tables as $table) {
-			if (strpos($table->getName(), '.')) {
-				list(, $tableName) = explode('.', $table->getName());
-			} else {
-				$tableName = $table->getName();
-			}
-			$this->emitCheckStep($tableName, $step++, count($tables));
-			// don't need to check for new tables
-			if (array_search($tableName, $existingTables) !== false) {
-				$this->checkTableMigrate($table);
-			}
-		}
-	}
-
-	/**
 	 * Create a unique name for the temporary table
 	 *
 	 * @param string $name
@@ -133,37 +105,6 @@ class Migrator {
 	 */
 	protected function generateTemporaryTableName($name) {
 		return $this->config->getSystemValue('dbtableprefix', 'oc_') . $name . '_' . $this->random->generate(13, ISecureRandom::CHAR_LOWER . ISecureRandom::CHAR_DIGITS);
-	}
-
-	/**
-	 * Check the migration of a table on a copy so we can detect errors before messing with the real table
-	 *
-	 * @param \Doctrine\DBAL\Schema\Table $table
-	 * @throws \OC\DB\MigrationException
-	 */
-	protected function checkTableMigrate(Table $table) {
-		$name = $table->getName();
-		$tmpName = $this->generateTemporaryTableName($name);
-
-		$this->copyTable($name, $tmpName);
-
-		//create the migration schema for the temporary table
-		$tmpTable = $this->renameTableSchema($table, $tmpName);
-		$schemaConfig = new SchemaConfig();
-		$schemaConfig->setName($this->connection->getDatabase());
-		$schema = new Schema([$tmpTable], [], $schemaConfig);
-
-		try {
-			$this->applySchema($schema);
-			$this->dropTable($tmpName);
-		} catch (DBALException $e) {
-			// pgsql needs to commit it's failed transaction before doing anything else
-			if ($this->connection->isTransactionActive()) {
-				$this->connection->commit();
-			}
-			$this->dropTable($tmpName);
-			throw new MigrationException($table->getName(), $e->getMessage());
-		}
 	}
 
 	/**

--- a/lib/private/DB/NoCheckMigrator.php
+++ b/lib/private/DB/NoCheckMigrator.php
@@ -30,9 +30,4 @@ use Doctrine\DBAL\Schema\Schema;
  * @package OC\DB
  */
 class NoCheckMigrator extends Migrator {
-	/**
-	 * @param \Doctrine\DBAL\Schema\Schema $targetSchema
-	 * @throws \OC\DB\MigrationException
-	 */
-	public function checkMigrate(Schema $targetSchema) {}
 }

--- a/lib/private/DB/OracleMigrator.php
+++ b/lib/private/DB/OracleMigrator.php
@@ -30,6 +30,35 @@ use Doctrine\DBAL\Schema\Schema;
 use Doctrine\DBAL\Schema\Table;
 
 class OracleMigrator extends NoCheckMigrator {
+
+	/**
+	 * Quote a column's name but changing the name requires recreating
+	 * the column instance and copying over all properties.
+	 *
+	 * @param Column $column old column
+	 * @param string $newName new name
+	 * @return Column new column instance with new name
+	 */
+	protected function quoteColumn($column) {
+		$newColumn = new Column(
+			$this->connection->quoteIdentifier($column->getName()),
+			$column->getType()
+		);
+		$newColumn->setAutoincrement($column->getAutoincrement());
+		$newColumn->setColumnDefinition($column->getColumnDefinition());
+		$newColumn->setComment($column->getComment());
+		$newColumn->setDefault($column->getDefault());
+		$newColumn->setFixed($column->getFixed());
+		$newColumn->setLength($column->getLength());
+		$newColumn->setNotnull($column->getNotnull());
+		$newColumn->setPrecision($column->getPrecision());
+		$newColumn->setScale($column->getScale());
+		$newColumn->setUnsigned($column->getUnsigned());
+		$newColumn->setPlatformOptions($column->getPlatformOptions());
+		$newColumn->setCustomSchemaOptions($column->getPlatformOptions());
+		return $newColumn;
+	}
+
 	/**
 	 * @param Schema $targetSchema
 	 * @param \Doctrine\DBAL\Connection $connection
@@ -43,23 +72,7 @@ class OracleMigrator extends NoCheckMigrator {
 			return new Table(
 				$this->connection->quoteIdentifier($table->getName()),
 				array_map(function(Column $column) {
-					$newColumn = new Column(
-						$this->connection->quoteIdentifier($column->getName()),
-						$column->getType()
-					);
-					$newColumn->setAutoincrement($column->getAutoincrement());
-					$newColumn->setColumnDefinition($column->getColumnDefinition());
-					$newColumn->setComment($column->getComment());
-					$newColumn->setDefault($column->getDefault());
-					$newColumn->setFixed($column->getFixed());
-					$newColumn->setLength($column->getLength());
-					$newColumn->setNotnull($column->getNotnull());
-					$newColumn->setPrecision($column->getPrecision());
-					$newColumn->setScale($column->getScale());
-					$newColumn->setUnsigned($column->getUnsigned());
-					$newColumn->setPlatformOptions($column->getPlatformOptions());
-					$newColumn->setCustomSchemaOptions($column->getPlatformOptions());
-					return $newColumn;
+					return $this->quoteColumn($column); 
 				}, $table->getColumns()),
 				array_map(function(Index $index) {
 					return new Index(
@@ -100,6 +113,9 @@ class OracleMigrator extends NoCheckMigrator {
 			$tableDiff->changedColumns = array_filter($tableDiff->changedColumns, function (ColumnDiff $column) {
 				return count($column->changedProperties) > 0;
 			});
+			$tableDiff->addedColumns = array_map(function(Column $column) {
+				return $this->quoteColumn($column);
+			}, $tableDiff->addedColumns);
 		}
 
 		return $schemaDiff;

--- a/lib/private/DB/SQLiteMigrator.php
+++ b/lib/private/DB/SQLiteMigrator.php
@@ -30,43 +30,6 @@ use Doctrine\DBAL\Types\BigIntType;
 use Doctrine\DBAL\Types\Type;
 
 class SQLiteMigrator extends Migrator {
-
-	/**
-	 * @param \Doctrine\DBAL\Schema\Schema $targetSchema
-	 * @throws \OC\DB\MigrationException
-	 *
-	 * For sqlite we simple make a copy of the entire database, and test the migration on that
-	 */
-	public function checkMigrate(\Doctrine\DBAL\Schema\Schema $targetSchema) {
-		$dbFile = $this->connection->getDatabase();
-		$tmpFile = $this->buildTempDatabase();
-		copy($dbFile, $tmpFile);
-
-		$connectionParams = [
-			'path' => $tmpFile,
-			'driver' => 'pdo_sqlite',
-		];
-		$conn = \Doctrine\DBAL\DriverManager::getConnection($connectionParams);
-		try {
-			$this->applySchema($targetSchema, $conn);
-			$conn->close();
-			unlink($tmpFile);
-		} catch (DBALException $e) {
-			$conn->close();
-			unlink($tmpFile);
-			throw new MigrationException('', $e->getMessage());
-		}
-	}
-
-	/**
-	 * @return string
-	 */
-	private function buildTempDatabase() {
-		$dataDir = $this->config->getSystemValue("datadirectory", \OC::$SERVERROOT . '/data');
-		$tmpFile = uniqid("oc_");
-		return "$dataDir/$tmpFile.db";
-	}
-
 	/**
 	 * @param Schema $targetSchema
 	 * @param \Doctrine\DBAL\Connection $connection

--- a/tests/lib/DB/MigratorTest.php
+++ b/tests/lib/DB/MigratorTest.php
@@ -102,8 +102,8 @@ class MigratorTest extends \Test\TestCase {
 		$this->connection->insert($this->tableName, ['id' => 2, 'name' => 'bar']);
 		$this->connection->insert($this->tableName, ['id' => 2, 'name' => 'qwerty']);
 
-		$migrator->checkMigrate($endSchema);
-		$this->fail('checkMigrate should have failed');
+		$migrator->migrate($endSchema);
+		$this->fail('migrate should have failed');
 	}
 
 	public function testUpgrade() {
@@ -115,7 +115,6 @@ class MigratorTest extends \Test\TestCase {
 		$this->connection->insert($this->tableName, ['id' => 2, 'name' => 'bar']);
 		$this->connection->insert($this->tableName, ['id' => 3, 'name' => 'qwerty']);
 
-		$migrator->checkMigrate($endSchema);
 		$migrator->migrate($endSchema);
 		$this->assertTrue(true);
 	}
@@ -134,7 +133,6 @@ class MigratorTest extends \Test\TestCase {
 		$this->connection->insert($this->tableName, ['id' => 2, 'name' => 'bar']);
 		$this->connection->insert($this->tableName, ['id' => 3, 'name' => 'qwerty']);
 
-		$migrator->checkMigrate($endSchema);
 		$migrator->migrate($endSchema);
 		$this->assertTrue(true);
 
@@ -173,7 +171,6 @@ class MigratorTest extends \Test\TestCase {
 		$migrator = $this->manager->getMigrator();
 		$migrator->migrate($startSchema);
 
-		$migrator->checkMigrate($endSchema);
 		$migrator->migrate($endSchema);
 
 		$this->assertTrue(true);
@@ -195,7 +192,6 @@ class MigratorTest extends \Test\TestCase {
 		$migrator = $this->manager->getMigrator();
 		$migrator->migrate($startSchema);
 
-		$migrator->checkMigrate($endSchema);
 		$migrator->migrate($endSchema);
 
 		$this->assertTrue(true);

--- a/tests/lib/DB/MigratorTest.php
+++ b/tests/lib/DB/MigratorTest.php
@@ -97,9 +97,6 @@ class MigratorTest extends \Test\TestCase {
 		return ($this->connection->getDatabasePlatform() instanceof OraclePlatform);
 	}
 
-	/**
-	 * @expectedException Doctrine\DBAL\Exception\UniqueConstraintViolationException
-	 */
 	public function testDuplicateKeyUpgrade() {
 		if ($this->isSQLite()) {
 			$this->markTestSkipped('sqlite does not throw errors when creating a new key on existing data');
@@ -115,8 +112,16 @@ class MigratorTest extends \Test\TestCase {
 		$this->connection->insert($this->tableName, ['id' => 2, 'name' => 'bar']);
 		$this->connection->insert($this->tableName, ['id' => 2, 'name' => 'qwerty']);
 
-		$migrator->migrate($endSchema);
-		$this->fail('migrate should have failed');
+		$caught = false;
+		try {
+			$migrator->migrate($endSchema);
+			$this->fail('migrate should have failed');
+		} catch (\Doctrine\DBAL\Exception\UniqueConstraintViolationException $e) {
+			$caught = true;
+			// makes PostgreSQL happier after an exception (and we don't have rollback yet on the public API...)
+			$this->connection->commit();
+		}
+		$this->assertTrue($caught);
 	}
 
 	public function testUpgrade() {

--- a/tests/lib/DB/MigratorTest.php
+++ b/tests/lib/DB/MigratorTest.php
@@ -46,16 +46,22 @@ class MigratorTest extends \Test\TestCase {
 
 		$this->config = \OC::$server->getConfig();
 		$this->connection = \OC::$server->getDatabaseConnection();
-		if ($this->connection->getDatabasePlatform() instanceof OraclePlatform) {
-			$this->markTestSkipped('DB migration tests are not supported on OCI');
-		}
 		$this->manager = new \OC\DB\MDB2SchemaManager($this->connection);
 		$this->tableName = strtolower($this->getUniqueID($this->config->getSystemValue('dbtableprefix', 'oc_') . 'test_'));
 	}
 
 	protected function tearDown() {
-		$this->connection->exec('DROP TABLE ' . $this->tableName);
+		$this->connection->exec('DROP TABLE ' . $this->connection->quoteIdentifier($this->tableName));
 		parent::tearDown();
+	}
+
+	private function getIndexName($tableName, $indexName) {
+		$indexName = $tableName . '_' . $indexName;
+		if ($this->isOracle()) {
+			// Oracle doesn't like long names...
+			return 'i' . substr(md5($indexName), 0, 29);
+		}
+		return $indexName;
 	}
 
 	/**
@@ -66,13 +72,13 @@ class MigratorTest extends \Test\TestCase {
 		$table = $startSchema->createTable($this->tableName);
 		$table->addColumn('id', 'integer');
 		$table->addColumn('name', 'string');
-		$table->addIndex(['id'], $this->tableName . '_id');
+		$table->addIndex(['id'], $this->getIndexName($this->tableName, 'id'));
 
 		$endSchema = new Schema([], [], $this->getSchemaConfig());
 		$table = $endSchema->createTable($this->tableName);
 		$table->addColumn('id', 'integer');
 		$table->addColumn('name', 'string');
-		$table->addUniqueIndex(['id'], $this->tableName . '_id');
+		$table->addUniqueIndex(['id'], $this->getIndexName($this->tableName, 'id'));
 
 		return [$startSchema, $endSchema];
 	}
@@ -87,12 +93,19 @@ class MigratorTest extends \Test\TestCase {
 		return $this->connection->getDriver() instanceof \Doctrine\DBAL\Driver\PDOSqlite\Driver;
 	}
 
+	private function isOracle() {
+		return ($this->connection->getDatabasePlatform() instanceof OraclePlatform);
+	}
+
 	/**
-	 * @expectedException \OC\DB\MigrationException
+	 * @expectedException Doctrine\DBAL\Exception\UniqueConstraintViolationException
 	 */
 	public function testDuplicateKeyUpgrade() {
 		if ($this->isSQLite()) {
 			$this->markTestSkipped('sqlite does not throw errors when creating a new key on existing data');
+		}
+		if ($this->isOracle()) {
+			$this->markTestSkipped('Does not work yet with Oracle, needs fixing index quoting');
 		}
 		list($startSchema, $endSchema) = $this->getDuplicateKeySchemas();
 		$migrator = $this->manager->getMigrator();
@@ -107,6 +120,9 @@ class MigratorTest extends \Test\TestCase {
 	}
 
 	public function testUpgrade() {
+		if ($this->isOracle()) {
+			$this->markTestSkipped('Does not work yet with Oracle, needs fixing index quoting');
+		}
 		list($startSchema, $endSchema) = $this->getDuplicateKeySchemas();
 		$migrator = $this->manager->getMigrator();
 		$migrator->migrate($startSchema);
@@ -120,6 +136,9 @@ class MigratorTest extends \Test\TestCase {
 	}
 
 	public function testUpgradeDifferentPrefix() {
+		if ($this->isOracle()) {
+			$this->markTestSkipped('Does not work yet with Oracle, needs fixing index quoting');
+		}
 		$oldTablePrefix = $this->config->getSystemValue('dbtableprefix', 'oc_');
 
 		$this->config->setSystemValue('dbtableprefix', 'ownc_');
@@ -140,6 +159,9 @@ class MigratorTest extends \Test\TestCase {
 	}
 
 	public function testInsertAfterUpgrade() {
+		if ($this->isOracle()) {
+			$this->markTestSkipped('Does not work yet with Oracle, needs fixing index quoting');
+		}
 		list($startSchema, $endSchema) = $this->getDuplicateKeySchemas();
 		$migrator = $this->manager->getMigrator();
 		$migrator->migrate($startSchema);
@@ -157,6 +179,9 @@ class MigratorTest extends \Test\TestCase {
 	}
 
 	public function testAddingPrimaryKeyWithAutoIncrement() {
+		if ($this->isOracle()) {
+			$this->markTestSkipped('Does not work yet with Oracle, needs fixing index quoting');
+		}
 		$startSchema = new Schema([], [], $this->getSchemaConfig());
 		$table = $startSchema->createTable($this->tableName);
 		$table->addColumn('id', 'integer');
@@ -172,6 +197,34 @@ class MigratorTest extends \Test\TestCase {
 		$migrator->migrate($startSchema);
 
 		$migrator->migrate($endSchema);
+
+		$this->assertTrue(true);
+	}
+
+	public function testAddingColumn() {
+		$schema = new Schema([], [], $this->getSchemaConfig());
+		$table = $schema->createTable($this->tableName);
+		$table->addColumn('id', 'integer');
+		$table->addColumn('name', 'string');
+
+		$migrator = $this->manager->getMigrator();
+		$migrator->migrate($schema);
+
+		$table->addColumn('newcolumn', 'string', [
+			'notnull' => false
+		]);
+
+		$migrator->migrate($schema);
+
+		$schemaManager = $this->connection->getSchemaManager();
+		$actualSchema = $schemaManager->createSchema();
+
+		// Oracle might change casing if double quotes were missing, so verify
+		// that the column names still match
+		$table = $actualSchema->getTable($this->tableName);
+		$this->assertEquals('id', $table->getColumn('id')->getName());
+		$this->assertEquals('name', $table->getColumn('name')->getName());
+		$this->assertEquals('newcolumn', $table->getColumn('newcolumn')->getName());
 
 		$this->assertTrue(true);
 	}

